Passing additional metadata to external signers

usmansaleem opened the pull request:
## PR Description
Pass additional meta-data to external signers so that they can apply slashing protection.

Removed MessageSignerService. Signer interface is now implemented directly by LocalSigner and ExternalSigner. This makes related metadata available to external signer's sign methods directly.

The Path (which is suffixed to external signer url) has been updated from `/signer/sign` to `/api/v1/eth2/sign`.

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Passing additional metadata to external signer

 -- so that external signer can apply slash protection rules

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Merge remote-tracking branch 'upstream/master' into external_signer_with_slash_protection

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Fixing LocalMessageSignerServiceTest

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Fixing SignerTest

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Merge remote-tracking branch 'upstream/master' into external_signer_with_slash_protection

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Merge remote-tracking branch 'upstream/master' into external_signer_with_slash_protection

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
MessageSignerService sign method signatures to take additionalProperties

Signed-off-by: Usman Saleem <usman@usmans.info>

ajsutton commented on ethereum/core/src/main/java/tech/pegasys/teku/core/signatures/LocalMessageSignerService.java:
```diff
@@ -43,22 +43,26 @@ public LocalMessageSignerService(final BLSKeyPair keypair, final AsyncRunner asy
   @Override
-  public SafeFuture<BLSSignature> signAggregationSlot(final Bytes signingRoot) {
+  public SafeFuture<BLSSignature> signAggregationSlot(
+      final Bytes signingRoot, final Map<String, Object> additionalProperties) {
```
I'm not sure this is the right approach.  We shouldn't need to create a map of properties when we already have the required data in a suitable format.  I'd suggest that the `MessageSignerService` interface is removed and just move everything up to the `Signer` interface which has the required data and is where the built-in slashing protection hooks in.

May need to extract a helper class to compute the signing root for data rather than having to duplicate that logic but that's doable.

usmansaleem replied:
👍

usmansaleem replied:
The primary purpose for this map was to serialize it as json to external signing service. If we delete MessageSignerService, that would mean LocalSigner and ExternalSigner extends Signer, is that correct assumption?

ajsutton replied:
Yep.

ajsutton replied:
Well, they'd implement `Signer` since its an interface but same idea. :)

usmansaleem replied:
MessageSignerService has been removed. Refactored signingRoot calculation in a util class. Signer is now implemented directly by LocalSigner and ExternalSigner. Do we still want to wrap externalSigner in SlashingProtectedSigner (in ValidatorLoader)?

ajsutton replied:
Yes definitely by default, but (possibly in a follow up PR) it would be good to have an option to disable the built-in slashing protection for external signers (it should not be possible to disable it for local validators).

usmansaleem replied:
👍  Thank you, that make sense.

usmansaleem committed:
Fixing LocalMessageSignerServiceTest

Signed-off-by: Usman Saleem <usman@usmans.info>

jframe commented on ethereum/core/src/main/java/tech/pegasys/teku/core/signatures/LocalMessageSignerService.java:
```diff
@@ -30,32 +31,38 @@ public LocalMessageSignerService(final BLSKeyPair keypair, final AsyncRunner asy
 
   @Override
-  public SafeFuture<BLSSignature> signAggregateAndProof(final Bytes signingRoot) {
+  public SafeFuture<BLSSignature> signAggregateAndProof(
```
This and the below sign methods don't require any additional parameters for the signing protection it's only the attestation and block

usmansaleem replied:
fixed in ExternalSigner.

usmansaleem committed:
Merge remote-tracking branch 'upstream/master' into external_signer_with_slash_protection

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Review suggestion - remove MessageSignerService

 -- LocalSigner and ExternalSigner implements Signer interface
-- Update references and test cases

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Merge remote-tracking branch 'upstream/master' into external_signer_with_slash_protection

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
minor refactoring signingrequestbody

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
cleaning SigningRequestBody

Signed-off-by: Usman Saleem <usman@usmans.info>

jframe commented on ethereum/core/src/test/java/tech/pegasys/teku/core/signatures/LocalSignerTest.java:
```diff
@@ -0,0 +1,121 @@
+  private final LocalSigner signer = new LocalSigner(KEYPAIR, asyncRunner);
+
+  @Test
+  public void shouldSignBlock1() {
```
This name seems odd, shouldSignBlock?

usmansaleem replied:
copied over from now deleted "SignerTest". Let me fix it.

jframe commented on ethereum/core/src/testFixtures/java/tech/pegasys/teku/core/signatures/TestSigner.java:
```diff
@@ -0,0 +1,83 @@
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+public class TestSigner implements Signer {
```
Is this needed? It looks almost the same as the LocalSigner.

usmansaleem replied:
LocalSigner uses asyncRunner. This one doesn't. May be I can modify LocalSigner slightly so that it can sign both sync and async mode. Then this can be deleted.

ajsutton replied:
Just use `LocalSigner` with `SyncAsyncRunner` like we do in all the other tests.

usmansaleem replied:
Thanks for this hint @ajsutton, I was missing this bit :).

jframe commented on validator/client/src/main/java/tech/pegasys/teku/validator/client/signer/ExternalSigner.java:
```diff
@@ -0,0 +1,162 @@
+
+  @Override
+  public SafeFuture<BLSSignature> createRandaoReveal(final UInt64 epoch, final ForkInfo forkInfo) {
+    return sign(signingRootForRandaoReveal(epoch, forkInfo), emptyMap());
```
can we include the type on all the sign requests. although there will be no additional data it would be nice for our metrics to seperate the type of signing request.

usmansaleem replied:
will do.

usmansaleem replied:
done

usmansaleem committed:
Removing TestSigner as LocalSigner can be used instead

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Send sign type to external signer

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Using SyncRunner in tests

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Merge remote-tracking branch 'upstream/master' into external_signer_with_slash_protection

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
Adding test fixture dependency

Signed-off-by: Usman Saleem <usman@usmans.info>

ajsutton commented on validator/client/src/main/java/tech/pegasys/teku/validator/client/signer/SigningRequestBody.java:
```diff
@@ -13,28 +13,51 @@
-  @JsonCreator
-  public SigningRequestBody(@JsonProperty("data") final String data) {
-    this.data = Bytes.fromHexString(data);
+  @JsonAnySetter private final Map<String, Object> metadata = new LinkedHashMap<>();
```
Any particular reason to used a `LinkedHashMap` instead of just `HashMap` here?

usmansaleem replied:
Not for any particular performance reason I recall, I can switch it back to HashMap.

ajsutton replied:
Simpler is better. :). The only reason I can think of is if you're asserting the order of the content somehow in tests.  As long as you're using `.equals` comparison of two `HashMap` instances they won't care about order.

usmansaleem replied:
The tests are asserting only outgoing json data in integration test, so using HashMap is working there.

ajsutton reviewed (approved):
LGTM.

usmansaleem committed:
Merge remote-tracking branch 'upstream/master' into external_signer_with_slash_protection

Signed-off-by: Usman Saleem <usman@usmans.info>

usmansaleem committed:
review suggestion - using Hashmap in SigningRequestBody

Signed-off-by: Usman Saleem <usman@usmans.info>
